File: python/wordy/wordy.py
def calculate(question):
  # remove 'what is', 'by', and trailing '?'
  query = question[:-1].replace(' by', '').split(' ')[2:]

  # validate expression length
  if len(query) < 3 and len(query) != 1: 
      raise ValueError('Invalid expression')

  # capture first operand
  res = int(query[0])

  # while there is more than just a digit in query
  # perform operation and concatenate result with remaining expression
  while len(query) > 1:
    if query[1] == 'plus':
      res += int(query[2])
    elif query[1] == 'minus':
      res -= int(query[2])
    elif query[1] == 'multiplied':
      res *= int(query[2])
    elif query[1] == 'divided':
      res /= int(query[2])
    else:
      raise ValueError('Invalid expression')

    query = [str(res)] + query[3:]
  
  return res

File: typescript/wordy/wordy.ts
class ArgumentError extends Error {}

type Operation = 'plus' | 'minus' | 'multiplied' | 'divided'

const OPERATIONS = {
  plus: (a: number, b: number): number =>  a + b,
  minus: (a: number, b: number): number =>  a - b,
  multiplied: (a: number, b: number): number =>  a * b,
  divided: (a: number, b: number): number =>  a / b,
}

const isOperation = (token: string): token is Operation =>
  token === 'plus' || token === 'minus' || token === 'multiplied' || token === 'divided'

class WordProblem {
  constructor(private readonly phrase: string) {}

  private getTokens = (): string[] =>
    this.phrase
      .slice(0, -1)
      .split(' ')
      .slice(2)
      .filter(token => token !== 'by')

  private getResult(a: string | number, operation: string, b: string) {
    if (!isOperation(operation)) {
      throw new ArgumentError()
    }
    return OPERATIONS[operation](Number(a), Number(b))
  }

  answer = (): number => {
    const [firstOperand, ...tokens] = this.getTokens()
    let res = (Number(firstOperand));
    while (tokens.length > 0) {
      const [operation, operand] = tokens.splice(0, 2)
      res = this.getResult(res, operation, operand)
    }
    return res
  }
}

export { WordProblem, ArgumentError }